### PR TITLE
Add `RequiredDeep` type

### PR DESCRIPTION
**Pull request by jarpoole**
Closes https://github.com/sindresorhus/type-fest/issues/273

This PR is based on a past attempt (see https://github.com/sindresorhus/type-fest/pull/401) with all the feedback integrated. The only exception is that I did not change `args` to `arguments` because as was helpfully noted this is an error inside a strict context like a module.

### CODE REVIEW

**Comment by jarpoole** [comment]
CI fails because of https://github.com/xojs/xo/issues/718. Seems like this was fixed in https://github.com/sindresorhus/eslint-plugin-unicorn/issues/2076. I don't really understand what is going wrong here...

File: source/required-deep.d.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import {BuiltIns} from './internal';
+
+type ExcludeUndefined<T> = Exclude<T, undefined>;
+
+/**
+Create a type from another type with all keys and nested keys set to required.
+
```
Comment (sindresorhus) [inline source/required-deep.d.ts:7]: Missing use-cases.

File: source/required-deep.d.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+type RequiredSettings = RequiredDeep<Settings>;
+Result: {
+	textEditor: {
+		fontSize: number;
+		fontColor: string;
+		fontWeight: number;
+	}
+	autocomplete: boolean;
+	autosave: boolean;
+}
```
Comment (sindresorhus) [inline source/required-deep.d.ts:31]: This should be a proper code comment

**Comment by sindresorhus** [comment]
Is this case handled? https://github.com/sindresorhus/type-fest/pull/401#issuecomment-1169144534

File: source/required-deep.d.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+export type RequiredDeep<T, E extends ExcludeUndefined<T> = ExcludeUndefined<T>> = E extends BuiltIns
+	? E
+	: E extends Map<infer KeyType, infer ValueType>
+		? Map<RequiredDeep<KeyType>, RequiredDeep<ValueType>>
+		: E extends Set<infer ItemType>
+			? Set<RequiredDeep<ItemType>>
+			: E extends ReadonlyMap<infer KeyType, infer ValueType>
+				? ReadonlyMap<RequiredDeep<KeyType>, RequiredDeep<ValueType>>
+				: E extends ReadonlySet<infer ItemType>
+					? ReadonlySet<RequiredDeep<ItemType>>
```
Comment (sindresorhus) [inline source/required-deep.d.ts:49]: Looks like `WeakSet` and `WeakMap` handling is missing.

**Comment by sindresorhus** [comment]
Should it handle promises? I looked at some different implementations of this kind of type and noticed this one handles promises: https://github.com/ts-essentials/ts-essentials/blob/master/lib/deep-required/index.ts

File: source/required-deep.d.ts (hunk 3 above)
Comment (jarpoole) [reply source/required-deep.d.ts:53]: Done!

File: source/required-deep.d.ts (hunk 1 above)
Comment (jarpoole) [reply source/required-deep.d.ts:7]: Done!

**Comment by jarpoole** [comment]
> Should it handle properties that are promises? I looked at some different implementations of this kind of type and noticed this one handles promises: https://github.com/ts-essentials/ts-essentials/blob/master/lib/deep-required/index.ts

Yes! I have added support.

File: source/required-deep.d.ts (hunk 2 above)
Comment (jarpoole) [reply source/required-deep.d.ts]: Good call!

**Comment by jarpoole** [comment]
> Is this case handled? [#401 (comment)](https://github.com/sindresorhus/type-fest/pull/401#issuecomment-1169144534)

I *think* I've handled that case now but I'm not 100% sure.

File: source/required-deep.d.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+/**
+Test if the given function has multiple call signatures.
+
+Needed to handle the case of a single call signature with properties.
+
+Multiple call signatures cannot currently be supported due to a TypeScript limitation.
+@see https://github.com/microsoft/TypeScript/issues/29732
+*/
+type HasMultipleCallSignatures<T extends (...arguments: any[]) => unknown> =
```
Comment (sindresorhus) [inline source/required-deep.d.ts:88]: Use the type in `source/internal.d.ts`.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Thanks :)

**Pull request merged**